### ISSUE TITLE
Typography Tools: Tracking

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#34345]
Tools:
- Font
- Size
- Line Height
- Appearance/Weight/Style
- Letter Spacing
- Letter Case

Mockup:
<img width="279" alt="Screen Shot 2021-09-24 at 2 00 42 PM" src="https://user-images.githubusercontent.com/25516611/134670861-ee2736d9-17a8-466c-8c49-b92b7725b88e.png">

### Components

- [x] Iterate on `ToolsPanel`
  - [x] #32392
  - [x] #34257
  - [x] #34157
  - [x] #34085
  - [x] #34632
  - [x] #34530
  - [x] #34726
  - [x] #34107
  - [x] #34028
  - [x] #35375
  - [x] #34725
  - [x] #35057
  - [x] #35058
  - [x] Switch to using `Grid` components for layout
  - [ ] #38059
- [x] Use `ToolsPanel` to render typography tools.
  - [x] Fix display of existing font size and line-height controls when in `ToolsPanel` ( PR: [#35451](https://github.com/WordPress/gutenberg/pull/35451) )
  - [x] https://github.com/WordPress/gutenberg/issues/35604
- [x] #35293
- [ ] Package an exportable, reusable `TypographyPanel` component using the components that are already available (similarly to what done [for the `DimensionsPanel`](https://github.com/WordPress/gutenberg/pull/32392)). Use this component as a "living reference" for iterating on each component
- [ ] Font size controls:
  - [x] When picking a font size from a list of predefined sizes, show either a ToggleGroup control (< 6 sizes) or a dropdown control (>= 6 sizes) (see screenshots below and this [early WIP prototype](https://g2-git-try-global-styles-sidebar-v2-itsjonq.vercel.app/iframe.html?id=examples-wip-fontsizepickerv2--default&viewMode=story))
    - [x] `ToggleGroupControl`: asses if it fulfills all design/UX requirements
    - [ ] `CustomSelectControl` may be used for the custom dropdown
      - [x] Fix initialValue computation when not passing an explicit value (PR #34490)
      - [x] Convert to TS + Emotion?
      - [ ] Add unit tests?
      - [x] Asses current look vs design (the needed design may not currently be possible, since the contents of each select option can currently only be specified as a string of text)
      - [ ] Asses API design (e.g. options passed as a prop vs `CustomSelectControlOption` component)?
  - [ ] When picking a custom font size (instead of picking for a list of predefs), show an input + slider
    - [x] #23908
    - [x] Asses if `FontSizePicker` (and therefore `UnitControl` and `RangeControl`) need any amends / design updates
  - [x] [Remove reset button](https://github.com/WordPress/gutenberg/pull/35395#issuecomment-937537987) and rely on `ToolsPanel` reset functionality
- [ ] `NumberControl`: 
  - [x] #34363 (related PR #35020)
  - [x] #34816 (related PR #34542)
  - [x] #33291 (related PRs: #34128 and #34566)
  - [ ] #34817
  - [ ] agree on next steps (e.g. update design)
- [x] `ToggleGroupControl`
  - [x] #35381
  - [x] #35382
  - [x] #35383
- [x] `FontAppearanceControl`
  - [x] #35461
- [ ] Update styling
  - [x] Sub-project: #36230
  - [x] #42000 
  - [x] #40049
  - [x] Update spin buttons in `NumberControl`
  - [x] #42782
  - [ ] Update `COLORS.ui.border`? (The mockups specify #bbb, a color that seems to be nonexistent in the current codebase. Unclear how this ties into the existing [color scheme](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/colors-values.js))

_When using predefined font-size value, the font size controls change depending on how many values need to be displayed_

| Predef (< 6 sizes) | Predef (>= 6 sizes) | Custom font size |
|---|---|---|
| ![image](https://user-images.githubusercontent.com/548849/131104825-5a3616bd-1b08-4ad4-bbe6-627791e1e7ee.png) | ![image](https://user-images.githubusercontent.com/548849/131104872-a61430ba-ce2b-4bd4-b7ea-f4b8045ad2f8.png) | ![image](https://user-images.githubusercontent.com/1083581/131821448-c1f3d0be-843a-456d-926a-aac3dd7b6936.png) |

### Integration

- [x] Ensure added features (like drop cap) are added at the bottom of the panel.
- [x] #32577
- [x] #23767
- [x] #20796
- [x] Replace usages of typography panel across blocks and make it consistent:
  - [x] #34381

### Design

- [x] Decide on how letter casing is to be presented.
- [x] Decide on a good name for the "appearance" control (weight, style).
- [x] Consider whether responsive options can be exposed in the UI: #24480

**Comment by carolinan** [comment]
The letter spacing control is currently not working: https://github.com/WordPress/gutenberg/issues/32577

**Comment by aaronrobertshaw** [comment]
> The letter spacing control is currently not working: #32577

I noticed this while testing changes removing `custom` prefixes from theme.json properties that didn't need it.

~A fix is included in that PR: https://github.com/WordPress/gutenberg/pull/34485~

Fixed in https://github.com/WordPress/gutenberg/pull/34515

**TL;DR** When the typography tools were grouped the letter-spacing support key wasn't updated as per the others.

**Comment by pablohoneyhoney** [comment]
Here's an update on some of the typography components and behaviors. 

<img width="279" alt="Screen Shot 2021-09-24 at 2 00 42 PM" src="https://user-images.githubusercontent.com/25516611/134670861-ee2736d9-17a8-466c-8c49-b92b7725b88e.png">

In particular regarding the Size component. Below there's a bit more detail around aspects not considered before, like variants on number of values (segmented UI for 4 or 5 values, and dropdown for >5), types of values if we want to consider t-shirt, tooltip to contextualize abstracted values, and toggling to input custom values.

<img width="636" alt="Size-component" src="https://user-images.githubusercontent.com/25516611/134670712-1b7278bc-dabc-4758-9848-5b8d5eebd6ef.png">

Relevant to #34574

**Comment by ntsekouras** [comment]
👋 - I've started looking at the font size picker and my first issue that is missing the designs is the `default` value and the `reset` button. How do these two fit in the control? -cc @pablohoneyhoney

**Comment by ciampo** [comment]
> 👋 - I've started looking at the font size picker and my first issue that is missing the designs is the `default` value and the `reset` button. How do these two fit in the control? -cc @pablohoneyhoney

It would be great if you could coordinate with @mirka , she's also been looking at this.

**Comment by ntsekouras** [comment]
Regarding `Letter spacing` control that [seemed important](https://github.com/WordPress/gutenberg/pull/31118#issuecomment-832545921) to have units, do we need to revisit something design-wise? I'm not sure if we have a control with `+/-` symbols and the `unit` select..

**Comment by javierarce** [comment]
@ntsekouras: 

> 👋 - I've started looking at the font size picker and my first issue that is missing the designs is the default value and the reset button. How do these two fit in the control? -cc @pablohoneyhoney

Is there an expectation to be able to reset every control or just the font size one? 

I think something like the following solution could work for components that span the entire column (in this example, 24 is the default value)

![image](https://user-images.githubusercontent.com/4933/136037753-774eac14-83eb-4c59-90c0-bcfdf5e8fd26.png)

**Comment by mtias** [comment]
The resetting comes from the tools panel and allows resetting each item. You can see it in this video:

https://github.com/WordPress/gutenberg/pull/33744

**Comment by javierarce** [comment]
> The resetting comes from the tools panel and allows resetting each item. You can see it in this video:

Oh, thanks, that makes sense… showing reset buttons for several components would make panels look very busy.

**Comment by mirka** [comment]
@pablohoneyhoney Are there public Figma links for these mockups, by the way? I see that the [components](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=5410%3A3723) are available, but I'm wondering if there's another mockup file I don't know about.

**Comment by mtias** [comment]
What's the status of this one?

**Comment by ciampo** [comment]
The status of the tracking issue seems quite up to date, meaning that we'd mostly need to refine/polish `FontSizePicker` and a few dependent components. The components squad hasn't had much capacity to move this forward, but we've been collaborating with folks like @ntsekouras @aaronrobertshaw @andrewserong and @ramonjd on PRs touching these components.

I also know that @mirka has worked/is working on aligning the design and behaviour of some of these components closer together.

**Comment by ramonjd** [comment]
> Consider whether responsive options can be exposed in the UI: https://github.com/WordPress/gutenberg/issues/24480

#### Update: _Wednesday, March 23 2022_

I've endeavoured to dive into a backend option for fluid typography based on existing `fontSizes` presets in theme.json.

- https://github.com/WordPress/gutenberg/pull/39529

When fluid typography is "activated" in theme.json, we generate `font-size` values using `clamp()`. 

The resulting font sizes can be applied to any text block in the editor UI, and the preset CSS vars to any root element or block in theme.json.

It's still work in progress as I look at various approaches, and consider the constraints on such functionality for an initial version, taking into consideration future development and (future) backwards compatibility should we change clamp algorithms.

The intention is to make fluid typography an available option for themes initially, so that themes can switch seamlessly between the two leaving no CSS artefacts behind. 🤞 

An [editor-based exploration](https://github.com/WordPress/gutenberg/pull/33543) is also underway.

**Comment by cbirdsong** [comment]
> > > ![image](https://user-images.githubusercontent.com/4933/136037753-774eac14-83eb-4c59-90c0-bcfdf5e8fd26.png)
> 
> > The resetting comes from the tools panel and allows resetting each item. You can see it in this video:
> 
> Oh, thanks, that makes sense… showing reset buttons for several components would make panels look very busy.

I think conspicuous reset buttons would actually be useful since it allows you to know at a glance what values you've adjusted away from the default. Here is how Webflow lets you know something has been modified and lets you reset to the default:
![CleanShot 2022-06-16 at 08 24 18](https://user-images.githubusercontent.com/1672206/174079763-a0c243c7-2f5d-4567-a5f7-fc86928fba10.gif)

**Comment by marceloaof** [comment]
Feedback after extensive use of the current typography tools.

One enhancement that I did not find a ticket for it, and that it would be appreciated to be developed, as it currently makes site editing more hard-working, is the possibility to:

- adjust the 4 default font sizes (--wp--preset--font-size--small, --wp--preset--font-size--medium, --wp--preset--font-size--large, --wp--preset--font-size--x-large)
- adjust theme font sizes
- and add custom font sizes
directly in the global styles sidebar.

Just as we can: edit WordPress default solid colors and gradients, edit theme colors, and add new colors in the global styles sidebar.

At the moment, this is only possible by editing theme.json. This solution brings problems.

Problem 1:
When font sizes are added via theme.json as custom values, they are not accessible in the dropdown control.
Example from the theme Twenty Twenty-Two:
		"custom": {
			"typography": {
				"font-size": {
					"huge": "clamp(2.25rem, 4vw, 2.75rem)",
					"gigantic": "clamp(2.75rem, 6vw, 3.25rem)",
					"colossal": "clamp(3.25rem, 8vw, 6.25rem)"
				}
		}
These font sizes (huge, gigantic and colossal) are added as --custom--typography--font-size-{name} variables, but they are not accessible in the dropdown control nor editable in the global styles sidebar.

Problem 2:
If these font sizes were to be added in theme.json as if they were official sizes, for example: --wp--preset--font-size--huge,
"fontSizes": [
				{
					"name": "Small",
					"size": "1rem",
					"slug": "small"
				},
then we would be able to select these font sizes in the dropdown control, but we still can't edit them directly in the global styles sidebar. And now the markup is wrong in the sense that this is not a preset font size.

**Comment by ramonjd** [comment]
### Update on fluid typography June 22, 2022

- https://github.com/WordPress/gutenberg/pull/39529

The PR to add fluid font sizes to theme.json has been [updated](
https://github.com/WordPress/gutenberg/pull/39529#issuecomment-1158545503) to support passing single values to the fluid type callback, and returning `clamp()` values based on a set of defaults: `minFontSizeFactor,` `maxFontSizeFactor` and `scaleFactor`.

This was the intention behind https://github.com/WordPress/gutenberg/pull/33543, which was testing converting custom font sizes to `clamp()` values.

I think it's a pretty good place to retest and hopefully get in!

**Comment by ramonjd** [comment]
### Update on fluid typography August 4, 2022

The theme.json fluid font size API and a couple of follow ups have been merged 🎉 

- https://github.com/WordPress/gutenberg/pull/39529
- https://github.com/WordPress/gutenberg/pull/42688
- https://github.com/WordPress/gutenberg/pull/42757

A follow up PR to add extra configuration to the fluid settings is underway:

- https://github.com/WordPress/gutenberg/pull/42489

**Comment by jasmussen** [comment]
For the segmented control of preset font sizes, we ended up going with numbers as a starting point:

<img width="271" alt="Screenshot 2022-08-03 at 10 22 13" src="https://user-images.githubusercontent.com/1204802/182560450-4bd3dc2e-9a41-4837-855b-2748f7f4934b.png">

However I'm finding myself missing the t-shirt sizes [as originally conceived](https://github.com/WordPress/gutenberg/issues/34345#issuecomment-926577197). They seem much more descriptive and universal after all:
<img width="274" alt="Screenshot 2022-08-03 at 10 24 45" src="https://user-images.githubusercontent.com/1204802/182560963-0d5b83d0-e068-4345-8217-5b3697d22475.png">

I would love to work with someone to try going back to those. It seems like it could be a small PR to get a feel for it, and I'd love to collaborate with someone to try this. @ramonjd or @andrewserong would you have bandwidth to pair with me on this one?

**Comment by mtias** [comment]
Let's try it, should be just an appearance update with the system we have in place. (Note we should handle XXL as well, even if not included in the default set.)

**Comment by ramonjd** [comment]
> would you have bandwidth to pair with me on this one?

I should have time to take a stab at it next week. 👍

**Comment by ramonjd** [comment]
@jasmussen Just in case I can't get to it next week, is there any further context (or better, an issue) in case someone else picks it up?

I think numbers are extracted from the theme.json size values (strips the units and uses the number values). So the `2` here is really `2rem` in my theme.json:

<img width="275" alt="Screen Shot 2022-08-05 at 3 30 46 pm" src="https://user-images.githubusercontent.com/6458278/183008082-60e6665c-9b74-4947-a208-d1c111481cf6.png">

It would be interesting to hear how we'd map t-shirt sizes to arbitrary values, unless we just assume that the order of font sizes in theme.json will be from smallest to largest? The first thing that comes to mind is a sorting function before we print them.

**Comment by mtias** [comment]
@ramonjd aren't the slugs already mapped behind the scenes anyways?

**Comment by ramonjd** [comment]
> aren't the slugs already mapped behind the scenes anyways?

@mtias If I'm interpreting things correctly, yes, the slugs that appear in `setings.typography.fontSizes[].slug` are mapped to the JS objects we use to build the fontsize controls. 

I'm still catching up with all the feedback over at:
- https://github.com/WordPress/gutenberg/pull/37038#pullrequestreview-822539765 and @jasmussen's [comment](https://github.com/WordPress/gutenberg/pull/37038#pullrequestreview-822539765)

Here's an example option object that's used in the editor, where the `name` property contains the value of the font size's slug:

```json
{
    "key": "medium",
    "value": "20px",
    "label": "20",
    "name": "Medium"
}
```

The font size picker parses the values and use the [integers as the labels](https://github.com/WordPress/gutenberg/blob/aac3e86b3858143c967bfcc28ab11f72d3aafa1b/packages/components/src/font-size-picker/utils.js#L94). So, yeah, we'd need to map the t-shirt labels before the controls are rendered I believe. 

Maybe update the titles as well. Something similar to what @ntsekouras was doing in this [commit](https://github.com/WordPress/gutenberg/pull/37038/commits/9a83492afd6361879ab430f186249db44700cc96).

It would be good to get an idea of the expectations. Here's a very basic implementation that labels the first 5 font sizes with t-shirt sizes, regardless of their values.

<details>

<summary>Font size picker with t-shirt sizes</summary>

https://user-images.githubusercontent.com/6458278/183323896-49924d98-a9c1-4ecb-b311-a74888607a23.mp4

</details>

**Comment by jasmussen** [comment]
> It would be good to get an idea of the expectations. Here's a very basic implementation that labels the first 5 font sizes with t-shirt sizes, regardless of their values.

Understanding that you could desync the custom labels from the t-shirt sizes, yes, that basic implementation, in my opinion, is exactly what I was thinking of. Worth putting in a PR and getting a feel for it.

My impression is that we shouldn't optimize towards mapping the t-shirt sizes for what a label might or might not be — mapping them sequentially, just like the numbers were, allows authors to map these on their own, and regardless as soon as you get more than 5 options, it becomes a dropdown.

**Comment by noisysocks** [comment]
Just doing a quick audit of how the Typography section in Global Styles looks compared to [the design](https://www.figma.com/file/oEkcAyhIvPFMVEAO8EImvA/Global-Styles?node-id=2920%3A66093). I'll then try to fix some of this up.

Designs:

| Panel | Font size variations |
| - | - |
| <img width="600" alt="Global styles typography design" src="https://user-images.githubusercontent.com/612155/185020644-89e143f1-8ee1-4c38-8725-78001fd4b0fb.png"> | <img width="636" alt="Size variations" src="https://user-images.githubusercontent.com/612155/185020659-3c07e7aa-e33d-4779-89da-dad28a7ded83.png"> |

What's in `trunk`:

| Font size radio | Font size dropdown | Custom font size |
| - | - | - |
| <img width="292" alt="Image" src="https://user-images.githubusercontent.com/612155/185020798-45e0a05d-8591-4d58-99d1-ff51d54be922.png"> | <img width="294" alt="Image" src="https://user-images.githubusercontent.com/612155/185020822-7fc477cb-98ae-4b7b-b5b4-5dddc6a09cd5.png"> | <img width="289" alt="Image" src="https://user-images.githubusercontent.com/612155/185020819-791e14b8-9bc7-433f-bca8-dd927e6271f9.png"> |

- _Heading level_
  - This isn't in the design. (That's ok)
  - ~Label should be uppercase and, imo, simply read _Heading level_~
  - ~Bottom margin too small~
  - [ ] Should be a `TabPanel` variation
- _Font_
  - [x] Label should read _Font_ not _Font family_ https://github.com/WordPress/gutenberg/pull/43304
  - ~Default font and system font should have the name of the font e.g. _SF Pro text - Primary_~ – not possible
- _Size_
  - [x] Too much whitespace between label and input https://github.com/WordPress/gutenberg/pull/43304
  - [x] Label hint should not show CSS unit (e.g. _(px)_) when a T-shirt size is selected https://github.com/WordPress/gutenberg/pull/44966
  - Dropdown
    - [x] Labels are missing from some values. (Very possible that I messed this up while testing) 
  - Custom
    - [x] Label hint should be _Custom_ (no parenthesis) https://github.com/WordPress/gutenberg/pull/44966
    - [x] Missing slider https://github.com/WordPress/gutenberg/pull/44598
    - [x] Has _Reset_ button https://github.com/WordPress/gutenberg/pull/35395#issuecomment-937537987 https://github.com/WordPress/gutenberg/pull/44180
    - [ ] Button containing "px" is too wide https://github.com/WordPress/gutenberg/pull/42000#issuecomment-1196629350, https://github.com/WordPress/gutenberg/issues/42879
- _Line height_
  - [x] Should use custom + / - buttons https://github.com/WordPress/gutenberg/pull/45333
- [x] Spacing between controls is uneven https://github.com/WordPress/gutenberg/pull/43304
- [x] _Line height_ and _Appearance_ should appear next to each other https://github.com/WordPress/gutenberg/pull/43304
- [x] _Letter spacing_ and _Letter case_ are missing. They should appear only in _Headings_ section https://github.com/WordPress/gutenberg/pull/43304

**Comment by noisysocks** [comment]
> * [ ]  _Letter spacing_ is missing
> * [ ]  _Letter case_ is missing

I'm trying to work out whether these two controls are supposed to be in Global Styles or not.

<img width="600" alt="Global styles typography design" src="https://user-images.githubusercontent.com/612155/185036105-1b594ef6-5804-4592-8658-5952dc2252cd.png">

I can see that _Letter spacing_ was added to block settings but not Global Styles in https://github.com/WordPress/gutenberg/pull/31118. Not sure if that was an intentional omission.

@aaronrobertshaw tells me that _Letter case_ used to exist in Global Styles but was removed because it was felt that this shouldn't be something that is controlled at such a high level. 

At any rate it would be good to revisit these decisions cc. @WordPress/gutenberg-design.

**Comment by mtias** [comment]
They should be available for headings in the global context

**Comment by mirka** [comment]
> * Heading level

As discussed in #42718, this heading level switcher needs to have `tabpanel` semantics. The easiest way would be to use the existing TabPanel component, but design-wise it won't fit seven items in a single row.

One possibility is to make a new TabPanel component (or style variant) that looks like a 32px height ToggleGroupControl — though there [hasn't been a confirmed decision or prioritization](https://github.com/WordPress/gutenberg/pull/42718#issuecomment-1210490709) on this yet from the design side. We can move forward with this if we can get confirmation that this is sufficient visual affordance for a user to distinguish between a `radio` and a `tabpanel`.

**Comment by jasmussen** [comment]
The tab-panel that looks like a segmented control, if it's super quick to build, seems like we should just knock it out! But with a freeze in a few weeks, of course it should have lower priority than anything that needs to go in 6.1.

**Comment by talldan** [comment]
> As discussed in https://github.com/WordPress/gutenberg/pull/42718, this heading level switcher needs to have tabpanel semantics. The easiest way would be to use the existing TabPanel component, but design-wise it won't fit seven items in a single row.

Has there been an accessibility check on making it a TabPanel? If it changes a value, my instinct would be that TabPanel is the wrong thing. Usually tabs don't have such an effect, they're just a way to navigate through information. It may need to be another component that looks like a TabPanel.

**Comment by mirka** [comment]
> If it changes a value, my instinct would be that TabPanel is the wrong thing.

In this case in the Global Styles (Typography ▸ Headings), it isn't changing a value, just flipping between views. (It _does_ look like a value selector on first sight, which is exactly why we need to differentiate 😄)

**Comment by jameskoster** [comment]
> which is exactly why we need to differentiate

I agree with this. If we have 'tabpanels that look like segmented controls' and/or 'segmented controls that look like tabs' then we risk creating an unpredictable UX. A style variation for tabs that is 'secondary' in terms of visual hierarchy might be good to explore. This would be useful in situations where we have nested tabs as well.

**Comment by mtias** [comment]
The semantic role of different heading levels is close to that of block variations:

<img width="337" alt="image" src="https://user-images.githubusercontent.com/548849/185575189-c774372c-99ce-45b2-97c5-148b8bca0560.png">

So maybe the issue is both visual and hierarchical, and heading level should go above the preview with more context around it.

**Comment by mirka** [comment]
> The semantic role of different heading levels is close to that of block variations

I don't think so. The block variation switcher behaves like a `radio` — its **value is persisted**, and the values of its **sub-controls are discarded** when you change the block variation to another one. The heading level switcher in GS is purely a `tabpanel` switcher — its **value is not persisted**, and the values of all their **sub-controls are persisted** regardless of which tabpanel you last see.

**Comment by mtias** [comment]
I mean the semantic role in terms of how they are represent from the block API perspective — we have registered heading levels as variations of headings — which means accessing them could be presented as separate blocks or the quick variations selection. This also goes for selecting the right heading element in the block inspector on a block instance.

**Comment by bobbingwide** [comment]
Re: #42782 I see that the code's been merged so I'm commenting here. 

The changes to make the labels uppercase have introduced a problem with Jetpack's Field consent block.

In the block editor, the text is all uppercased, which makes it difficult to edit.

![image](https://user-images.githubusercontent.com/2474435/191031452-a0c45aa0-732c-44d4-8742-b25bd3567a46.png)

See also https://github.com/bobbingwide/bobbingwide/issues/77

Elsewhere it's been suggested that using `text-transform: uppercase` is bad for accessibility. 
Is it too late to revisit this issue?

**Comment by mirka** [comment]
> Re: #42782 I see that the code's been merged so I'm commenting here.

Replied in https://github.com/WordPress/gutenberg/issues/42782#issuecomment-1251073331 👍

**Comment by noisysocks** [comment]
I've completed most of the items in https://github.com/WordPress/gutenberg/issues/34345#issuecomment-1217390408 and in the issue description. The remaining items are, I think, fairly marginal and not worth prioritising over https://github.com/WordPress/gutenberg/issues/33094 which is the key focus for the editor team.

Here's how the Typography panel now looks in `trunk`—things are very close to what's in Figma:

https://user-images.githubusercontent.com/612155/199615854-30124692-f194-4fc2-ae2c-136aa7c2ec74.mp4

I think we can close this overview issue out and use smaller bug/enhancement issues for any further iteration that's required.
